pcdsutils/tests/test_logging.py

```diff
@@ -151,3 +152,89 @@ def inner_test(filtered: bool):
     inner_test(filtered=True)
     warnings_filter.uninstall()
     inner_test(filtered=False)
+
+
+@pytest.fixture(scope='function')
+def callback_demoter():
+    filt = pcdsutils.log.OphydCallbackExceptionDemoter.install()
+    yield filt
+    filt.uninstall()
+
+
+def test_exception_filter(
+    callback_demoter: pcdsutils.log.OphydCallbackExceptionDemoter,
+    caplog: pytest.LogCaptureFixture,
+):
+    caplog.set_level(logging.DEBUG)
+
+    def zerodiv(*args, **kwargs):
+        1/0
+
+    sig = ophyd.Signal(name='sig')
+    sig.subscribe(zerodiv, run=False)
+
+    def inner_test(filtered: bool):
+        callback_demoter.reset_counter()
+        total_cnt = 10
+        for cnt in range(total_cnt):
+            caplog.clear()
+            sig.put(cnt)
+            target_records = [
+                rec for rec in caplog.records if rec.name == 'ophyd.objects'
+            ]
+            assert target_records, f"Did not find object log records! {cnt=}"
+            assert len(target_records) == 1, f"Too many records! {cnt=}"
+            record = target_records[0]
+            if not filtered or cnt == 0:
+                assert record.levelno == logging.ERROR, f"{filtered=}, {cnt=}"
```

Comment:
Having seen the syntaxerror that flake8 was raising on this file in our discussion last week, it gave a bit of extra time for me to think about the implications here.

While in the test suite, `pcdsutils.tests` is still part of `pcdsutils`. typhos (a non PCDS-specific package) relies on pcdsutils, and some other facilities are or may start to use it.

Given that, I think it'd be wise to backpedal a bit on 3.9+-only; pcdsdevices and hutch-python are fine, but less so our core libraries.

It's possible that downstream users just wouldn't care - or it's not such a big deal - but I want to raise this now before we get much further using 3.9+ features. Thoughts?

Reply:
Oh that's a good point, I didn't even consider the typhos/dependencies angle. How does a follow-up PR to backpedal to 3.7+ before any tagging sound?

Reply:
numpy drop schedule for reference:

Reply:
uh, this turned out to be really easy to pull back to 3.7- I'll scrap my new branch and just drop the commits into this one

Reply:
the future annotations import must be doing some magic